lib/ddtrace/contrib/rack/rum_injection.rb

```diff
@@ -0,0 +1,239 @@
+      # the application frontend.
+      # rubocop:disable Metrics/ClassLength
+      class RumInjection
+        include Datadog::Environment::Helpers
+
+        RUM_INJECTION_FLAG = 'datadog.rum_injection_flag'.freeze
+        INLINE = 'inline'.freeze
+        IDENTITY = 'identity'.freeze
+        HTML_CONTENT = 'text/html'.freeze
+        XHTML_CONTENT = 'application/xhtml+xml'.freeze
+        HEAD_TAG_OPEN = '<head'.freeze
+        TAG_CLOSE = '>'.freeze
+        NO_CACHE = 'no-cache'.freeze
+        NO_STORE = 'no-store'.freeze
+        PRIVATE = 'private'.freeze
+        MAX_AGE = 'max-age'.freeze
+        SMAX_AGE = 's-maxage'.freeze
+        MAX_AGE_ZERO = 'max-age=0'.freeze
+        SMAX_AGE_ZERO = 's-maxage=0'.freeze
+        CONTENT_TYPE_STREAMING = 'text/event-stream'.freeze
+        CACHE_CONTROL_HEADER = 'Cache-Control'.freeze
+        CONTENT_DISPOSITION_HEADER = 'Content-Disposition'.freeze
+        CONTENT_ENCODING_HEADER = 'Content-Encoding'.freeze
+        CONTENT_LENGTH_HEADER = 'Content-Length'.freeze
+        CONTENT_TYPE_HEADER = 'Content-Type'.freeze
+        EXPIRES_HEADER = 'Expires'.freeze
+        SURROGATE_CACHE_CONTROL_HEADER = 'Surrogate-Control'.freeze
+        TRANSFER_ENCODING_HEADER = 'Transfer-Encoding'.freeze
+        ACTION_CONTROLLER_INSTANCE = 'action_controller.instance'.freeze
+        TRANSFER_ENCODING_CHUNKED = 'chunked'.freeze
+
+        def initialize(app)
+          @app = app
+        end
+
+        def call(env)
+          result = @app.call(env)
+
+          begin
+            return result unless configuration[:rum_injection_enabled] == true && env['rack.hijack?'] != true
+
+            status, headers, body = result
+
+            # need significant safety here to ensure result is parsable + injectable
+            # shouldnt be gzipped/compressed yet since we've injected our middleware in the stack after rack deflater
+            # or any other compression middleware for that matter
+            # also ensure its non-cacheable, is html, is not streaming, and is not an attachment
+            injectable = should_inject?(headers, env, status)
+
+            if injectable
+              trace_id = current_trace_id
+
+              # do not inject if no trace or trace is not sampled
+              return result unless trace_id
+
+              # we need to insert the trace_id and expiry meta tags
+              unix_time = DateTime.now.strftime('%Q').to_i
+
+              html_comment = html_comment_template(trace_id, unix_time)
+
+              # update content length and return new response if we don't fail on rum injection
+              if body.respond_to?(:each)
+                # inject html comment into first available fragment and then end early so we do not
+                # iterate over entire response
+                body.each do |fragment|
+                  fragment.insert(0, html_comment)
+                  break
```

Comment:
`break` wouldn't work with the semantics of `#each`, as it would forever consume the first element:
```ruby
File.write('/tmp/test', "first\nsecond")
iter = File.open('/tmp/test').each_line
iter.each { break }
iter.each { |l| puts l }
#> second
```